File: DynamicProgramming_problems/MaximumTotalDamageWithSpellCasting.java

// Problem link - https://leetcode.com/problems/maximum-total-damage-with-spell-casting/?envType=daily-question&envId=2025-10-11

package DynamicProgramming_problems;

import java.util.*;

public class MaximumTotalDamageWithSpellCasting {
    // dp array for memoization
    static Long[] dp;

    // utility method to find lower bound
    private static int lowerBoundUtils(List<Long> nums, int left, int right, long target) {
        int result = right; // If not found, return right (out of bounds)
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums.get(mid) >= target) {
                result = mid;
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        
        return result;
    }

    // recursive method to take or skip the power and find maximum damage
    private static long solveRec(List<Long>nums, int i, int n, Map<Long, Long> frqMap){
        // base case: if index out of bounds, no more damage can be added
        if(i >= n){
            return 0;
        }

        // if this state is already computed, return cached result
        if(dp[i] != null){
            return dp[i];
        }

        // option 1 - skip the current power value, as it might fall in forbidden range
        long skipOption = solveRec(nums, i+1, n, frqMap);

        // option 2 - take current power value
        // find next power value that is not forbiddden range
        // using lower_bound to find first element >= nums.get(i) + 3
        int j = lowerBoundUtils(nums, i+1, n, nums.get(i) + 3);
        
        // total damage = (current power * its frequency) + damage from non-conflicting powers
        long takeOption = nums.get(i) * frqMap.get(nums.get(i)) + solveRec(nums, j, n, frqMap);

        // store and return the maximum of both options
        return dp[i] = Math.max(skipOption, takeOption);
    }

    public static long maximumTotalDamage(int[] power) {
        // step 1 - count frequency of each power value
        Map<Long, Long> frqMap = new HashMap<>();
        for(int p : power){
            long pLong = (long) p;
            frqMap.put(pLong, frqMap.getOrDefault(pLong, 0L) + 1);
        }

        // step 2 - extract unique power values
        List<Long> nums = new ArrayList<>(frqMap.keySet());

        // step 3 - sort in ascending order
        Collections.sort(nums);

        // step 4 - initialize dp array
        int n = nums.size();
        dp = new Long[n+1];

        // step 5 - solve recursively
        return solveRec(nums, 0, n, frqMap);
    }

    public static void main(String[] args) {
        // int[] power = {1,1,3,4};

        int[] power = {7,1,6,6};
        System.out.println(maximumTotalDamage(power));
    }
}

File: DynamicProgramming_problems/MaximumTotalDamageWithSpellCasting.js

// Problem link - https://leetcode.com/problems/maximum-total-damage-with-spell-casting/description/?envType=daily-question&envId=2025-10-11

let maximumTotalDamage = function(power) {
    // step 1 - count frequency of each power value
    const freqMap = new Map();
    for (const p of power) {
        freqMap.set(p, (freqMap.get(p) || 0) + 1);
    }
    
    // step 2 - extract and sort unique power values
    const nums = Array.from(freqMap.keys()).sort((a, b) => a - b);
    const n = nums.length;
    
    // step 3 - initialize memoization array
    const dp = new Array(n).fill(-1);
    
    // recursive function to find maximum damage
    const solveRec = (i) => {
        // base case: reached end
        if (i >= n) {
            return 0;
        }
        
        // return cached result
        if (dp[i] !== -1) {
            return dp[i];
        }
        
        // option 1: skip current power
        const skipOption = solveRec(i + 1);
        
        // option 2: take current power
        // find next non-conflicting power using binary search
        let left = i + 1;
        let right = n;
        const target = nums[i] + 3;
        
        // binary search for first element >= target
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (nums[mid] >= target) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        
        const takeOption = nums[i] * freqMap.get(nums[i]) + solveRec(left);
        
        // step 4 - store and return maximum
        dp[i] = Math.max(skipOption, takeOption);
        return dp[i];
    };
    
    // step 5 - start recursion from index 0
    return solveRec(0);
};

// const power = [1,1,3,4];

const power = [7,1,6,6];
console.log(maximumTotalDamage(power));